### ISSUE TITLE
SDL2: KEYUP events not returning unicode value

### ISSUE THREAD

**Issue by mcpalmer1980** [pygame/pygame#1236]
For some reason KEYDOWN events only return a unicode value some of the time. I know TEXTINPUT is meant to replace it, and that does work well, but this will cause many existing programs to fail. It took me a long time to notice it myself, as it only effects the text input dialog of my menu class, which is rarely use. It wasn't until I started working on new flyboy levels and needed to name an event that I noticed the problem.

It seems to me that KEYDOWN.unicode reports a character about 5% of the time a key is pressed. This seems wierd: one would expect an all or nothing result.

My text input dialog is complicated due to the many conditions handled, such as cursor position, deletion, key repeat, etc. I worked around this problem by ignoring the certatin keydown events if SDL2 is defined, and handling TEXTINPUT events if SDL2 is defined.

	def update_input(self, events, max_length):
		if events == False: return
		for event in events:
			if event.type == pygame.KEYDOWN:
				self.cursor_visible = True # So the user sees where he writes

				# If none exist, create counter for that key:
				if not event.key in self.keyrepeat_counters:
					self.keyrepeat_counters[event.key] = [0, event.unicode]

				if event.key == K_BACKSPACE: # FIXME: BACKSPACES NOT REPEATING IN SDL2
					self.input_string = self.input_string[:max(self.cursor_position - 1, 0)] + \
										self.input_string[self.cursor_position:]

					# Subtract one from cursor_pos, but do not go below zero:
					self.cursor_position = max(self.cursor_position - 1, 0)
				elif event.key == K_DELETE:
					self.input_string = self.input_string[:self.cursor_position] + \
										self.input_string[self.cursor_position + 1:]

				elif event.key == K_RETURN or event.key == K_KP_ENTER:
					self.keyrepeat_counters = {}
					return True

				elif event.key == K_RIGHT:
					# Add one to cursor_pos, but do not exceed len(input_string)
					self.cursor_position = min(self.cursor_position + 1, len(self.input_string))

				elif event.key == K_LEFT:
					# Subtract one from cursor_pos, but do not go below zero:
					self.cursor_position = max(self.cursor_position - 1, 0)

				elif event.key == K_END:
					self.cursor_position = len(self.input_string)

				elif event.key == K_HOME:
					self.cursor_position = 0

				# REGULAR KEY PRESSED, HANDLE IT AS KEYDOWN FOR SDL1
				# IGNORE IT IF SDL2 TO AVOID OCCASIONAL REDUNDANT INPUTS
				elif len(self.input_string) <= max_length and not SDL2:
					self.input_string = self.input_string[:self.cursor_position] + \
										event.unicode + \
										self.input_string[self.cursor_position:]
					self.cursor_position += len(event.unicode) # Some are empty, e.g. K_UP

			# GET CHARACTER FROM TEXTINPUT EVENT IF SDL2
			elif SDL2 and event.type == TEXTINPUT:
				if len(self.input_string) <= max_length:
					# If no special key is pressed, add unicode of key to input_string
					self.input_string = self.input_string[:self.cursor_position] + \
										event.text + \
										self.input_string[self.cursor_position:]
					self.cursor_position += len(event.text) # Some are empty, e.g. K_UP

			elif event.type == KEYUP:
				# *** Because KEYUP doesn't include event.unicode, this dict is stored in such a weird way
				if event.key in self.keyrepeat_counters:
					del self.keyrepeat_counters[event.key]

		# Update key counters:
		for key in self.keyrepeat_counters :
			self.keyrepeat_counters[key][0] += self.clock.get_time() # Update clock
			# Generate new key events if enough time has passed:
			if self.keyrepeat_counters[key][0] >= self.keyrepeat_intial_interval_ms:
				self.keyrepeat_counters[key][0] = self.keyrepeat_intial_interval_ms - \
													self.keyrepeat_interval_ms

				event_key, event_unicode = key, self.keyrepeat_counters[key][1]
				pygame.event.post(pygame.event.Event(KEYDOWN, key=event_key, unicode=event_unicode))

		# Rerender text surface:
		self.input_surface = self.body_font.render(self.input_string, 1, self.select_color)

		# Update self.cursor_visible
		self.cursor_ms_counter += self.clock.get_time()
		if self.cursor_ms_counter >= self.cursor_switch_ms:
			self.cursor_ms_counter %= self.cursor_switch_ms
			self.cursor_visible = not self.cursor_visible

		if self.cursor_visible:
			cursor_x_pos = self.body_font.size(self.input_string[:self.cursor_position])[0]
			# Without this, the cursor is invisible when self.cursor_position > 0:
			if self.cursor_position > 0:
				cursor_x_pos -= self.cursor_surface.get_width()
			self.input_surface.blit(self.cursor_surface, (cursor_x_pos, 0))
			self.cursor_off = cursor_x_pos + self.body_font.size('O')[0]

		#self.clock.tick()
		return False

**Related Docs:** https://www.pygame.org/docs/ref/event.html

**Comment by charlesej** [comment]
I was unable to use the code you included, so I tried the following code in an attempt to isolate the issue of unicode values not being reported on `KEYDOWN` events. I tried reproducing it with and without using the `set_repeat()` function, but was unable see any instances where pygame is not providing the correct unicode values (using pygame 2.0.0.dev3 with python 3.7.2 and 2.7.10). I did, however, uncover another issue which might be related. Issue #1238 indicates that some event types (e.g. `KEYDOWN` events, which your code above is posting) are not being posted to the event queue in pygame 2.0.0.dev3.

---

```python
import pygame

pygame.init()

screen = pygame.display.set_mode((400, 200))
pygame.display.flip()
clock = pygame.time.Clock()
running = True

#pygame.key.set_repeat(50, 50)

while running:
    clock.tick(20)

    for e in pygame.event.get():
        if e.type == pygame.QUIT:
            running = False
        elif e.type == pygame.KEYDOWN:
            print('KEYDOWN:', e.unicode, e.key)
            if e.key == pygame.K_ESCAPE:
                running = False

pygame.quit()
```

---

System details:
- os: windows 10 (64bit)
- python: 3.7.2 (64bit) and 2.7.10 (64bit)
- pygame: 2.0.0.dev3 (SDL: 2.0.10) at 7a5ed0b864645e7c325d80aa8866e837559c7a1c

**Comment by mcpalmer1980** [comment]
The lengthy code segment I included earlier was an example of how I needed to work around this problem in an existing game of nearly 10 thousand lines. It was never meant to exhibit the problem. Also, I failed to mention my platform as I have several times in the past. I apologize for the confusion and for wasting your time.

Platform: Linux Mint 18.3 Mate, python 3.5.2, pygame 2.0.0.dev3

Example Code:
	import pygame as pg

	pg.init()
	pg.display.set_mode((320,200))
	clock = pg.time.Clock()
	looping = True

	while looping:
		for e in pg.event.get():
			if e.type == pg.KEYDOWN:
				print(e.unicode or 'None')
				if e.key == pg.K_ESCAPE:
					looping = False
			if e.type == pg.TEXTINPUT:
				print(e.text)
		clock.tick(10)

Example Output:
	pygame 2.0.0.dev3 (SDL 2.0.9, python 3.5.2)
	Hello from the pygame community. https://www.pygame.org/contribute.html
	None
	f
	None
	d
	None
	s
	s
	e
	None
	j
	None
	k
	None
	k
	None

I received the same results in both python 2.7 and pygame 2.0.0.dev2. I never tested in Windows.

**Comment by dlon** [comment]
Does not reproduce:

OS: Windows 10
Python 3.7.2, 2.0.0.dev3
Python 3.7.2, 2.0.0.dev2
Python 2.7, 2.0.0.dev1

**Comment by mcpalmer1980** [comment]
so it must be platform specific. I'm running Linux Mint 18.3 Mate

**Comment by charlesej** [comment]
@mcpalmer1980 Thanks for following up with some example code and output.

**Comment by robertpfeiffer** [comment]
Yeah, I get the same thing on Ubuntu Linux 18.04. It's easy enough to work around, but for the sake of backward compatibility, this should be fixed.

**Comment by illume** [comment]
This fixed a bit better in #1398 

- [ ] it's broken when pressing 2-4 keys at once.
- [ ] check windows again

The TEXTINPUT/TEXTEDITING API is the way forward for getting text, since it handles things a lot better.

**Comment by illume** [comment]
It should probably be in the keyup event (as mentioned in #1524).

**Comment by illume** [comment]
This PR fixes it for me: https://github.com/pygame/pygame/pull/2398
`ankith26-event-get`

To reproduce run this example and press four keys at once. See the unicode field on the keydown events.
```python
python3 -m pygame.examples.eventlist
```
The branch also adds a unicode field to the key up events.

**Comment by illume** [comment]
Closing because it's fixed in #2398 thanks @ankith26
